Draft - TS Guidelines (#17104)
strapi/strapi: docs/docs/guides/03-typescript.md

title: Typescript
---

# Development Guidelines

This document provides guidelines for developing on the Strapi codebase using TypeScript. It covers various topics, including types location, generics, breaking changes, factories and more.

## 1. Type Location

When working with types on the Strapi codebase, it is essential to consider the location of types. These are the three common approaches and situations we document:

### a. Colocation

In the colocation approach, type definitions are colocated with the corresponding code. This means that types are defined in the same file or module where they are used.

**This approach is suitable when the types are specific to a particular module**.

> _Note: In some situation, colocated types can be exported too if needed by other modules_

Reviewer: ```suggestion
_Note: In some situations, colocated types can still be exported if needed by other modules_
```
But also, what are those situations?

Author: This for instance https://github.com/strapi/strapi/pull/17104/files#r1242366691

@joshuaellis I think props falls in this category

Author: 6913093336838e93042f925aa4fb2c82d0a7db1d

Author: 6913093336838e93042f925aa4fb2c82d0a7db1d

Example:

```typescript
// config.ts
interface Config {
  prop1: string;
  prop2: number;
}

function update(config: Config): Config {
  // ...
}
```

> ### b. Definitions Files

Author: Now that I think more about this (and taken some time to look at other libraries), I'm wondering whether having regular .ts file for type declaration wouldn't make more sense?
And then keeping .d.ts for module declaration & extension.

My only concern is that it adds some level of complexity + you'll sometimes end up with .d.ts next to .ts which might not be very clear.

What do you think @joshuaellis / @innerdvations / @alexandrebodin ?

Reviewer: In theory I think you're right. We should be using regular .ts files and then exporting whatever we explicitly want to export in each module in a .d.ts file. Pretty close to what we're already doing but renaming most of our .d.ts -> .ts

However, I don't think it hurts anything to keep them in .d.ts files, especially since we still have so much javascript and jsdoc usage in the codebase, and they tend to be picked up better by editors than regular .ts files. Plus .d.ts enforces that we ONLY include typings in those files and no code can accidentally slip in (although maybe there's a rule for that we could add for the types dir? no idea)

I don't have a strong opinion either way, curious to hear arguments from others though.

Reviewer 2: someone did a twitter thread on this – https://twitter.com/mattpocockuk/status/1672918127414333440 https://twitter.com/mattpocockuk/status/1673017324587171842

Author: Yes I saw this one, but I don't think it fits very well Strapi's context.
I mean, it basically ends up being the pros/cons that we've listed above :confused:

Reviewer: I'm leaning towards `.ts` at this point for the skipLibCheck, and that it seems to be best practices (after reading through the twitter debate and links there). I don't see any major pros to `.d.ts`

> **When types are shared across multiple files or modules, or when dealing with types that does not belong to a specific module**, it is recommended to use definition files (`.d.ts`) to separate type declarations from their usage.

Reviewer 2: Does this make sense in the context of frontend components? i.e. we may export props from one components to be used to create the props of another, it feels strange that these props wouldn't live with it's component?

Author: I think in the context of components it definitely make sense to keep the props colocated with the component

Example:

```typescript
// service.d.ts
export type Admin = Record<string, any>;

// uid.d.ts
export type Admin = `admin::${string}`;

// index.d.ts
export * as Service from './service';
export * as UID from './uid';

// src/index.ts
import type { Service, UID } from './index';

type AdminService = Record<UID.Admin, Service.Admin>;
```

### b. Imports

When importing **only types** from a module, `import type` must be used.

The `import type` syntax imports only the types and eliminates the runtime impact of importing the actual values. This helps reduce the bundle size and improves performance by avoiding unnecessary code execution.

Example:

```typescript
// ❌ don't

import { Attribute, Common } from '@strapi/strapi';

// ✅ do

import type { Attribute, Common } from '@strapi/strapi';
```

When importing **both types and implementations** from a module:

- `import type` must be used for types
- `import` must be used for the rest

Example:

```typescript
// ❌ don't

import { User, findUser } from './user';

const user: User = findUser();

// ✅ do

import type { User } from './user';
import { findUser } from './user';

const user: User = findUser();
```

### c. Exports

> In most cases, it is recommended to use regular exports (export) instead of the default export (export default). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.

Reviewer: ```suggestion
In most cases, it is recommended to use regular exports (`export`) instead of the default export (`export default`). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
```

Author: 3b8540879cf2f47138fe1855f8f141666d181f4a

> In most cases, it is recommended to use regular exports (export) instead of the default export (export default). Regular exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.

Reviewer 2: ```suggestion
In most cases, it is recommended to use named exports (export) instead of the default export (export default). Named exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.
```

I believe they're called "named exports"

Author: That make sense, thanks

However, there are a few exceptions where using export default might be appropriate. For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.

Using regular exports provides a clearer and more explicit interface for module consumers, promoting better code organization and maintainability. However, it is important to consider the context before deciding whether to use default exports.

> ## 3. Generic Types' Parameters

Reviewer: ```suggestion
## 3. Parameters for Generics
```

When working with generic types, it is required to follow naming conventions and use descriptive names for type parameters. This helps improve code readability and maintainability over the whole codebase.

- For simple types with a single self-explanatory parameter, it is allowed to **use `T`**
- For any other generic types' parameters, it is required to **use a meaningful PascalCased variable name prefixed by `T`**

Example:

```typescript
// ❌ don't
function query<T extends object, U extends keyof T>(data: T, properties: U[]): { [K in U]: T[K] } {
  // ...
}

// ✅ do
function query<TValue extends object, TKey extends keyof TValue>(
  data: TValue,
  properties: TKey[]
): { [TIter in TKey]: TValue[TIter] } {
  // ...
}

// ❌ avoid
type NonEmpty<TValue extends string> = TValue extends '' ? never : TValue;

// ✅ prefer

> type NonEmpty<T extends string> = T extends '' ? never : T;

Reviewer: I thought we always prefered verbose, we just had an exception to allow single letters if it was self-explanatory.

Author: Don't you consider this to be self-explanatory? (between the name of the type, the extends string & the fact that the type definition is quite straight forward)

Reviewer: (I swear I replied to this but now I don't see it)

I mean that I thought we had discussed both options would be ✅ rather than avoiding the first one. If someone wants to write it out, that's just as acceptable, but they don't need to if it's that simple.

```

Choose meaningful names for type parameters that accurately describe their purpose within the context of the function or type.

## 4. Avoiding Breaking Changes for Users

When making changes to the Strapi types, it is important to consider the impact on existing users. To avoid breaking changes, follow these guidelines:

> - **Avoid removing or changing public APIs:** Public types APIs include functions, classes, or interfaces/types exposed to users as part of the official/documented Strapi API. Removing or changing them without providing appropriate deprecations or migration paths can break users' existing code.

Reviewer: Not necessary but it would be nice if we could think of an example for how to do this when changing a type definition.

Author: I've added it for the second bullet point item 36ca09280f3d8d21609b3c66b8c424ed3885d5ca

> - **Use deprecation warnings instead:** If you need to deprecate an existing API, provide deprecation warnings and suggest alternative approaches. Give users a reasonable amount of time to adapt their code to the changes.

Reviewer: Same here, how would a type definition deprecation look for someone making that change?

Author: 36ca09280f3d8d21609b3c66b8c424ed3885d5ca

- **Release breaking changes in major versions:** Breaking changes should be released as part of major version updates (e.g., from v4.x.x to v5.x.x). Major versions indicate significant changes that may require code modifications from users.

By following these guidelines, you can minimize the impact on users' code when introducing changes to the Strapi types.

## 5. Using Shared Type Registries for Public APIs

> When dealing with public oriented APIs, it's recommended to use dynamic type registries in order to offer a pleasant developer experience to both maintainers and users.

Reviewer 2: I think this is potentially too backend orientated to be called "public apis", for example the helper-plugin is a public facing package with numerous public APIs – the props of a component for instance.

Author: The whole document is definitely too backend oriented. I'll work on that, thanks for the feedback :slightly_smiling_face:

Author: I've added mention of frontend topics such as the component's props, @joshuaellis don't hesitate to tell me if you see any other topic worth mentioning :eyes:

Reviewer 3: This would apply to the FE with customization of state management most likely.

This document provides guidelines for developing on the Strapi codebase using TypeScript. It covers various topics, including types location, generics, breaking changes, factories and more.

## 1. Type Location

When working with types on the Strapi codebase, it is essential to consider the location of types. These are the three common approaches and situations we document:

### a. Colocation

In the colocation approach, type definitions are colocated with the corresponding code. This means that types are defined in the same file or module where they are used.

**This approach is suitable when the types are specific to a particular module**.

_Note: In some situations, colocated types can still be exported if needed by other modules (e.g. React Component Props)_

Example:

```typescript
// config.ts

interface Config {
  prop1: string;
  prop2: number;
}

function update(config: Config): Config {
  // ...
}
```

```typescript
// Box.tsx

import * as React from 'react';

export interface TaskProps {
  name: string;
  priority: 'high' | 'medium' | 'low';
  due: Date;
}

export const TaskProps = ({ name, priority, due }: TaskProps) => {
  // ...
};
```

### b. Definitions Files

> **When types are shared across multiple files or modules, or when dealing with types that does not belong to a specific module**, it is recommended to use definition files (`.d.ts`) to separate type declarations from their usage.

Reviewer 3: 🤔  do you have resource explaining the pros&cons of using definition files instead of let's say a `types.ts` file ? for shared types in a folder ?

Reviewer 3: Well yup +1

Definitions files provide a centralized location for types, making it easier to maintain and reuse them.

Example:

```typescript
// service.d.ts
interface CommonService {
  [key: keyof any]: unknown;
}

// users.ts
const usersService: CommonService = {
  // ...
};

// books.ts
const booksService: CommonService = {
  // ...
};
```

### c. Utils

**For type utilities used across different packages**, it is mandatory to place them in their related module's definition file (.d.ts).

- Strapi utils **must** be placed in `@strapi/strapi`'s types module
- Packages' specific utils **must** be placed in their respective types' module

Basically, utility modules contain reusable type definitions that can be imported wherever needed. This approach promotes code reusability and separation of concerns.

Example:

> ```typescript

Reviewer 3: Kind of same question with the ts migration why not just `types/core/uid.ts` ?

// core/uid.d.ts
export type Admin = `admin::${string}`;

// admin.ts
import type { UID } from '@strapi/strapi';

function getService(uid: UID.Admin): object {
  // ...
}
```

Choose the appropriate approach based on the size and complexity of your feature and the level of reusability of the types.

## 2. Modules, Imports and Exports

### a. Modules

In TypeScript, modules can be used to organize and encapsulate related types and code. Modules provide a way to mimic namespaces and avoid polluting the global scope with numerous type declarations. By grouping related types within modules, it is possible to improve code organization and make it easier to navigate and understand the codebase.

This is especially true when working with packages exports or utilities, where the exported components needs to be scoped.

Example:

```typescript
// service.d.ts
export type Admin = Record<string, any>;

// uid.d.ts
export type Admin = `admin::${string}`;

// index.d.ts
export * as Service from './service';
export * as UID from './uid';

// src/index.ts
import type { Service, UID } from './index';

type AdminService = Record<UID.Admin, Service.Admin>;
```

### b. Imports

> When importing **only types** from a module, `import type` must be used.

Reviewer 3: Will this be enforce with eslint ? If so I would remove them from the guidelines so it only includes things we haven't a lint rule for :)

Author: What about keeping this + adding a section about eslint + adding backlinks between both?

Author: Also, 2 things:
- It's possible to enforce that with an eslint rule
- It's also possible to have a mix that preserves both: `import { type Article, createArticle } from 'smth';`

Reviewer 3: Great news. If we ever change the rule the guidelines won't be up to date. It's really hard to maintain those kind of things that less is better IMO. I would instead add a section that link to the strapi commong eslint config and the local config file to know about all the others rules

The `import type` syntax imports only the types and eliminates the runtime impact of importing the actual values. This helps reduce the bundle size and improves performance by avoiding unnecessary code execution.

Example:

```typescript
// ❌ don't

import { Attribute, Common } from '@strapi/strapi';

// ✅ do

import type { Attribute, Common } from '@strapi/strapi';
```

When importing **both types and implementations** from a module:

- `import type` must be used for types
- `import` must be used for the rest

Example:

```typescript
// ❌ don't

import { User, findUser } from './user';

const user: User = findUser();

// ✅ do

import type { User } from './user';
import { findUser } from './user';

const user: User = findUser();
```

> ### c. Exports

Reviewer 3: I agree with what's said but that might just be too vague to be in the guideline 🤔 we probably need a specific list for this

Author: What kind of list are you thinking about?

It's basically: always use export {} except in a few situations like "`For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.`"

Reviewer 3: Yes these are just examples :) we need clear rules with clear exceptions :)

Author: I've tried to make it clearer, not sure if this is what you meant. Waiting for your feedback

Author: [eff4e40](https://github.com/strapi/strapi/pull/17104/commits/eff4e403216736b51cf3ef9bc4fd281446adfe5e)

Reviewer 3: Should be fine for now

In most cases, it is recommended to use named exports (`export`) instead of the default export (`export default`). Named exports provide more flexibility and clarity when importing and using modules. They allow importing specific members of a module explicitly, which improves code readability and maintainability.

However, there are a few exceptions where using export default might be appropriate. For example, when exporting a single primary entity or class from a module, or when working with modules that have a natural default export, such as a configuration object.

Using regular exports provides a clearer and more explicit interface for module consumers, promoting better code organization and maintainability. However, it is important to consider the context before deciding whether to use default exports.

## 3. Parameters for Generics

When working with generic types, it is required to follow naming conventions and use descriptive names for type parameters. This helps improve code readability and maintainability over the whole codebase.

> - For simple types with a single self-explanatory parameter, it is allowed to **use `T`**

Reviewer 3: Could we say "For generic type declaration with a single parameter" 🤔  so we make it clearer we only allow it for pure ts declaration and not inside code.

Same question as above what about eslint ?

Author: > and not inside code.

What about

```ts
function doSomething<T extends string>(value: T): T {
  // ...
}
```

It's code, but I don't think it makes sense to name the generic parameter

- For any other generic types' parameters, it is required to **use a meaningful PascalCased variable name prefixed by `T`**

Example:

```typescript
// ❌ don't
function query<T extends object, U extends keyof T>(data: T, properties: U[]): { [K in U]: T[K] } {
  // ...
}

// ✅ do
function query<TValue extends object, TKey extends keyof TValue>(
  data: TValue,
  properties: TKey[]
): { [TIter in TKey]: TValue[TIter] } {
  // ...
}

// ✅ ok
type NonEmpty<TValue extends string> = TValue extends '' ? never : TValue;

// ✅ prefer
type NonEmpty<T extends string> = T extends '' ? never : T;
```

Choose meaningful names for type parameters that accurately describe their purpose within the context of the function or type.

## 4. Avoiding Breaking Changes for Users

> 

Reviewer 3: We can add a callout to verify if a type you change is exported by the package. (I'm wondering how much we could automate that eventually, there might be lib that will check that)

When making changes to the Strapi types, it is important to consider the impact on existing users. To avoid breaking changes, follow these guidelines:

- **Avoid removing or changing public APIs:**

  Public types APIs include functions, classes, or interfaces/types exposed to users as part of the official/documented Strapi API. Removing or changing them without providing appropriate deprecations or migration paths can break users' existing code.

  This also applies to public React components (through their props).

- **Use deprecation warnings instead:**

>   If you need to deprecate an existing type, provide deprecation warnings and suggest alternative approaches. Give users a reasonable amount of time to adapt their code to the changes.

Reviewer 3: Reasonable = next major :D. We need to maintain a retro compat interface. 

The only exception is when we need to fix a bug
